### PR TITLE
HCX-666 : Java Integrator SDK - Notification APIs Feature support

### PR DESCRIPTION
**Pull request by AbhiGaddi**

### CODE REVIEW

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/interfaces/IncomingRequest.java [hunk 1]
```diff
@@ -166,6 +166,8 @@ public interface IncomingRequest {
      *
      * @return It is a boolean value to understand the final status is successful or failure.
      */
+
     boolean sendResponse(Map<String,Object> error, Map<String,Object> output) throws JsonProcessingException;
+    Map<String,Object> receiveNotification(String requestBody, Map<String,Object> output, Config config) throws Exception;
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/interfaces/IncomingRequest.java:171]: Add Java Documentation

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/interfaces/OutgoingRequest.java [hunk 2]
```diff
@@ -178,4 +179,6 @@ public interface OutgoingRequest {
      */
     boolean initializeHCXCall(String jwePayload, Operations operation, Map<String,Object> response, Config config) throws Exception;
 
+    boolean sendNotification(String topicCode, String recipientType, List<String> recipients, String message, Map<String,String> templateParams, String correlationID, Map<String,Object> output, Config config) throws Exception;
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/interfaces/OutgoingRequest.java:182]: Add Java Documentation

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java [hunk 3]
```diff
@@ -55,8 +55,10 @@ public class HCXOutgoingRequest extends FhirPayload implements OutgoingRequest {
 
     private static final Logger logger = LoggerFactory.getLogger(HCXOutgoingRequest.class);
 
+    NotificationService notificationService;
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java:58]: why this a public variable?

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXIncomingRequest.java [hunk 4]
```diff
@@ -104,11 +108,33 @@ public boolean sendResponse(Map<String,Object> error, Map<String,Object> output)
...
+    public Map<String, Object> receiveNotification(String jwsPayload, Map<String, Object> output, Config config) throws Exception {
+        jwsPayload = deserializePayload(jwsPayload);
+        Map<String,Object> requestPayload = JSONUtils.deserialize(jwsPayload,Map.class);
+        NotificationRequest notificationRequest = new NotificationRequest((String) requestPayload.get(Constants.PAYLOAD));
+        if (notificationRequest.getJwsPayload().isEmpty()) {
+            throw new ClientException("JWS Token cannot be empty");
+        }
+        Map<String, Object> headers = notificationRequest.getHeaders();
+        Map<String, Object> payload = notificationRequest.getPayload();
+        Map<String, Object> headersMap = (Map<String, Object>) headers.get(Constants.NOTIFICATION_HEADERS);
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXIncomingRequest.java:129]: get notification headers from notification request dto

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java [hunk 5]
```diff
@@ -149,19 +151,30 @@ public boolean encryptPayload(Map<String,Object> headers, String fhirPayload, Ma
...
-        boolean result = false;
-        if(status == 202){
-            result = true;
-            logger.info("Processing outgoing request has completed ::  response: {}", response.get(Constants.RESPONSE_OBJ));
-        } else {
-            logger.error("Error while processing the outgoing request :: status: {} :: response: {}", status, response.get(Constants.RESPONSE_OBJ));
-        }
+    public boolean initializeHCXCall(String jwePayload, Operations operation, Map<String, Object> response, Config config) throws Exception {
+        boolean result;
+        result = Utils.initializeHCXCall(jwePayload, operation, response, config);
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java:156]: why result variable is required? why cannot return directly Utils.initializeHCXCall.

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java [hunk 6]
```diff
@@ -149,19 +151,30 @@ public boolean encryptPayload(Map<String,Object> headers, String fhirPayload, Ma
...
+    @Override
+    public boolean sendNotification(String topicCode, String recipientType, List<String> recipients, String message, Map<String, String> templateParams, String correlationId, Map<String, Object> output, Config config) throws Exception {
+        try {
+            boolean result;
+            NotificationRequest notificationRequest = new NotificationRequest(topicCode, message, templateParams, recipientType, recipients, correlationId, config);
+            NotificationService.validateNotificationRequest(notificationRequest);
+            Map<String, Object> requestBody = NotificationService.createNotificationRequest(notificationRequest, output, message);
+            Map<String, Object> response = new HashMap<>();
+            result = initializeHCXCall(JSONUtils.serialize(requestBody), Operations.NOTIFICATION_NOTIFY, response, config);
+            output.putAll(response);
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java:169]: why creating new response map and pass to initializeHCXCall? pass output map to the method.

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/service/NotificationService.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        } else if (notificationRequest.getRecipientType().isEmpty()) {
+            throw new ClientException("Recipient type cannot be empty");
+        }
+    }
+
+    private static String getMessage(NotificationRequest notificationRequest,Map<String, Object> output, String message) throws Exception {
+        Map<String, Object> searchRequest = Map.of("filters", new HashMap<>());
+        if (!notificationRequest.getTemplateParams().isEmpty()) {
+            Map<String, Object> responseMap = new HashMap<>();
+            Utils.initializeHCXCall(JSONUtils.serialize(searchRequest), Operations.NOTIFICATION_LIST, responseMap, notificationRequest.getConfig());
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/service/NotificationService.java:35]: why creating new response map and pass to initializeHCXCall? pass output map to the method.

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/utils/Operations.java [hunk 8]
```diff
@@ -17,7 +17,9 @@ public enum Operations {
     COMMUNICATION_REQUEST("/communication/request","CommunicationRequest"),
     COMMUNICATION_ON_REQUEST("/communication/on_request","Bundle"),
     PREDETERMINATION_SUBMIT("/predetermination/submit","Bundle"),
-    PREDETERMINATION_ON_SUBMIT("/predetermination/on_submit","Bundle");
+    PREDETERMINATION_ON_SUBMIT("/predetermination/on_submit","Bundle"),
+    NOTIFICATION_NOTIFY("/notification/notify",""),
+    NOTIFICATION_LIST("/notification/topic/list","");
```
Comment (shiva-rakshith) [inline java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/utils/Operations.java:22]: Add EOB fetch and on fetch to the list.

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**